contracts/randdrop/src/contract.rs

```diff
@@ -0,0 +1,843 @@
+        msg: to_binary(&ProxyExecuteMsg::GetNextRandomness {
+            job_id: "randdrop-".to_string() + info.sender.as_str(),
+        })?,
+
+        funds: vec![config.nois_proxy.price], // Pay from the contract
+    });
+
+    // Register randdrop participant
+    let participant_data = &ParticipantData {
+        nois_randomness: None,
+        base_randdrop_amount: amount,
+        is_winner: None,
+    };
+    PARTICIPANTS.save(deps.storage, &info.sender, participant_data)?;
+
+    Ok(response)
+}
+
+pub fn execute_receive(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    callback: NoisCallback,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+
+    // callback should only be allowed to be called by the proxy contract
+    // otherwise anyone can cut the randomness workflow and cheat the randomness by sending the randomness directly to this contract
+    ensure_eq!(
+        info.sender,
+        config.nois_proxy.address,
+        ContractError::UnauthorizedReceive
+    );
+    let randomness: [u8; 32] = callback
+        .randomness
+        .to_array()
+        .map_err(|_| ContractError::InvalidRandomness)?;
+
+    // extract participant address
+    let job_id = callback.job_id;
+    let participant_address = job_id
+        .strip_prefix("randdrop-")
+        .unwrap_or_else(|| panic!("Strange, how is the job-id not prefixed with randdrop-"));
+    let participant_address = deps.api.addr_validate(participant_address)?;
+
+    // Make sure the participant is registered
+    let participant_data = PARTICIPANTS.load(deps.storage, &participant_address)?;
+    if participant_data.nois_randomness.is_some() || participant_data.is_winner.is_some() {
+        panic!("Strange, participant's randomness already received")
+    }
+    let mut msgs = Vec::<CosmosMsg>::new();
+
+    let randdrop_amount = match is_randdrop_winner(&participant_address, randomness) {
+        true => {
+            let randdrop_amount =
+                participant_data.base_randdrop_amount * Uint128::from(AIRDROP_ODDS);
+            msgs.push(
+                BankMsg::Send {
+                    to_address: participant_address.to_string(),
+                    amount: vec![Coin {
+                        amount: randdrop_amount,
+                        denom: config.randdrop_denom.clone(),
+                    }],
+                }
+                .into(),
+            );
+            randdrop_amount
+        }
+        false => Uint128::new(0),
+    };
+
+    // verify not claimed
+    if CLAIMED.has(deps.storage, &participant_address) {
+        panic!("Strange, participant's randdrop already claimed")
+    }
+
+    // Update claim
+    CLAIMED.save(deps.storage, &participant_address, &randdrop_amount)?;
+
+    // Delete Participant Data
+    PARTICIPANTS.remove(deps.storage, &participant_address);
```

Comment:
When we delete this data here, the is_winner query cannot work anymore. I'd keep the data here and update it.

Reply:
One of the nice features of merklized randdrop is that the contract is light in computation and in storage. 
If we have thousands of people participating without cleaning the storage we might end up with a huge state. what do you think ?

Reply:
Good point. I think storing the data here is fine since we only have entries for actual participants which is a much smaller group than the eligable addresses.

Right now, the PARTICIPANTS data is only available between requesting and receiving the randomness, 20-80 seconds. This is a bit pointless.

What about merging `PARTICIPANTS` and `CLAIMED` into one map and keep the entries? Then we have only one database entry per participant.

Reply:
yeah if we decide to keep the data then the natural choice would be to merge `PARTICIPANTS` and `CLAIMED` . 
The `PARTICIPANTS` serves the goal of conducting the randdrop so it is needed during the 20-80 seconds. then it's deleted because it's not needed anymore so it is not pointless. the `CLAIMED` allows you to know afterwards who won and who lost amongst participants.
This being said we can also just leave the state grow and not care so much.